// react-native-sheet/lib/commonjs/features/BottomSheet/BottomSheet.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BottomSheet = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _styles = require("./styles");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const BottomSheet = /*#__PURE__*/(0, _react.memo)( /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  const {
    height,
    colorScheme,
    onRequestClose,
    backdropClosesSheet = true,
    backdropBackgroundColor,
    sheetBackgroundColor,
    sheetStyle,
    draggable = true,
    showDragIcon = true,
    dragIconColor,
    dragIconStyle,
    borderRadius = 10,
    contentContainerStyle,
    onOpenStart,
    onOpenFinish,
    onCloseStart,
    onCloseFinish,
    openTime = 300,
    closeTime = 300,
    children
  } = props;
  const isDraggable = _reactNative.Platform.OS !== 'web' && !!draggable;
  const colorSchemeName = (0, _reactNative.useColorScheme)() || 'light';
  const isDarkMode = (0, _react.useMemo)(() => {
    // Color Scheme from props passed as 'auto' or undefined
    if (colorScheme === 'auto' || !colorScheme) {
      // Then use OS hook result
      return colorSchemeName === 'dark';
    } // Otherwise return user choice


    return colorScheme === 'dark';
  }, [colorScheme, colorSchemeName]);
  const [visible, setVisible] = (0, _react.useState)(false);
  const animatedHeight = (0, _react.useMemo)(() => new _reactNative.Animated.Value(0), []);
  const pan = (0, _react.useMemo)(() => new _reactNative.Animated.ValueXY(), []);
  const setModalVisibility = (0, _react.useCallback)(newVisible => {
    // Open modal
    if (newVisible) {
      setVisible(true);
      if (typeof onOpenStart === 'function') onOpenStart();

      _reactNative.Animated.timing(animatedHeight, {
        toValue: height,
        duration: openTime,
        useNativeDriver: false
      }).start(() => {
        if (typeof onOpenFinish === 'function') onOpenFinish();
      });

      return;
    } // Close modal


    if (typeof onCloseStart === 'function') onCloseStart();

    _reactNative.Animated.timing(animatedHeight, {
      toValue: 0,
      duration: closeTime,
      useNativeDriver: false
    }).start(() => {
      pan.setValue({
        x: 0,
        y: 0
      });
      setVisible(false);
      if (typeof onCloseFinish === 'function') onCloseFinish();
    });
  }, [animatedHeight, closeTime, height, onCloseFinish, onCloseStart, onOpenFinish, onOpenStart, openTime, pan]); // Respond to pan changes

  const panResponder = (0, _react.useRef)(_reactNative.PanResponder.create({
    onStartShouldSetPanResponder: () => true,
    onPanResponderMove: (e, gestureState) => {
      if (gestureState.dy > 0) {
        _reactNative.Animated.event([null, {
          dy: pan.y
        }], {
          useNativeDriver: false
        })(e, gestureState);
      }
    },
    onPanResponderRelease: (_e, gestureState) => {
      const gestureLimitArea = height / 3;
      const gestureDistance = gestureState.dy;

      if (gestureDistance > gestureLimitArea) {
        setModalVisibility(false);
      } else {
        _reactNative.Animated.spring(pan, {
          toValue: {
            x: 0,
            y: 0
          },
          useNativeDriver: false
        }).start();
      }
    }
  })).current; // If `height` prop changes while open, respond and change to new height immediately

  (0, _react.useEffect)(() => {
    if (visible) {
      _reactNative.Animated.timing(animatedHeight, {
        toValue: height,
        duration: 0,
        useNativeDriver: false
      }).start();
    } // eslint-disable-next-line react-hooks/exhaustive-deps

  }, [height]); // we only want the useEffect to run when height prop changes
  // Expose specific methods on hook

  (0, _react.useImperativeHandle)(ref, () => ({
    show: () => setModalVisibility(true),
    hide: () => setModalVisibility(false)
  })); // Return nothing if not visible

  if (!visible) return null;
  return /*#__PURE__*/_react.default.createElement(_reactNative.Modal, {
    transparent: true,
    visible: visible,
    onRequestClose: onRequestClose ? onRequestClose : () => setModalVisibility(false)
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [_styles.styles.wrapper, {
      backgroundColor: backdropBackgroundColor || `#25252599`,
      minHeight: height
    }]
  }, /*#__PURE__*/_react.default.createElement(_reactNative.TouchableOpacity, {
    testID: "backdrop",
    disabled: !backdropClosesSheet,
    style: _styles.styles.backdrop,
    activeOpacity: 1,
    onPress: () => setModalVisibility(false)
  }), /*#__PURE__*/_react.default.createElement(_reactNative.Animated.View, _extends({}, isDraggable && panResponder.panHandlers, {
    style: [{
      transform: pan.getTranslateTransform()
    }, _styles.styles.container, {
      height: animatedHeight,
      borderTopRightRadius: borderRadius,
      borderTopLeftRadius: borderRadius,
      backgroundColor: sheetBackgroundColor || isDarkMode ? '#222222' : '#F3F3F3'
    }, sheetStyle],
    testID: "animated-view"
  }), isDraggable && showDragIcon && /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: _styles.styles.draggableContainer
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    testID: "drag-icon",
    style: [_styles.styles.draggableIcon, dragIconStyle, {
      backgroundColor: dragIconColor || isDarkMode ? '#444444' : '#A3A3A3'
    }]
  })), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [{
      height
    }, contentContainerStyle]
  }, children))));
}));
exports.BottomSheet = BottomSheet;
